cmd/vela-server/schedule.go

```diff
@@ -0,0 +1,357 @@
+
+	b := new(library.Build)
+	b.SetAuthor(s.GetCreatedBy())
+	b.SetBranch(r.GetBranch())
+	b.SetClone(r.GetClone())
+	b.SetCommit(r.GetBranch())
+	b.SetDeploy(s.GetName())
+	b.SetEvent(constants.EventSchedule)
+	b.SetMessage(fmt.Sprintf("triggered for %s schedule with %s entry", s.GetName(), s.GetEntry()))
+	b.SetRef(fmt.Sprintf("refs/heads/%s", b.GetBranch()))
+	b.SetRepoID(r.GetID())
+	b.SetSender(s.GetUpdatedBy())
+	b.SetSource(fmt.Sprintf("%s/tree/%s", url, b.GetBranch()))
+	b.SetStatus(constants.StatusPending)
+	b.SetTitle(fmt.Sprintf("%s received from %s", constants.EventSchedule, url))
+
+	// populate the build link if a web address is provided
+	if len(metadata.Vela.WebAddress) > 0 {
+		b.SetLink(fmt.Sprintf("%s/%s/%d", metadata.Vela.WebAddress, r.GetFullName(), b.GetNumber()))
+	}
+
+	var (
+		// variable to store the raw pipeline configuration
+		config []byte
+		// variable to store executable pipeline
+		p *pipeline.Build
+		// variable to store pipeline configuration
+		pipeline *library.Pipeline
+		// variable to control number of times to retry processing pipeline
+		retryLimit = 5
+		// variable to store the pipeline type for the repository
+		pipelineType = r.GetPipelineType()
+	)
+
+	// implement a loop to process asynchronous operations with a retry limit
+	//
+	// Some operations taken during this workflow can lead to race conditions failing to successfully process
+	// the request. This logic ensures we attempt our best efforts to handle these cases gracefully.
+	for i := 0; i < retryLimit; i++ {
+		logrus.Debugf("compilation loop - attempt %d", i+1)
+		// check if we're on the first iteration of the loop
+		if i > 0 {
+			// incrementally sleep in between retries
+			time.Sleep(time.Duration(i) * time.Second)
+		}
+
+		// send API call to attempt to capture the pipeline
+		pipeline, err = database.GetPipelineForRepo(b.GetCommit(), r)
+		if err != nil { // assume the pipeline doesn't exist in the database yet
+			// send API call to capture the pipeline configuration file
+			config, err = scm.ConfigBackoff(u, r, b.GetCommit())
+			if err != nil {
+				return fmt.Errorf("unable to get pipeline config for %s/%s: %w", r.GetFullName(), b.GetCommit(), err)
+			}
+		} else {
+			config = pipeline.GetData()
+		}
+
+		// send API call to capture repo for the counter (grabbing repo again to ensure counter is correct)
+		r, err = database.GetRepoForOrg(r.GetOrg(), r.GetName())
+		if err != nil {
+			err = fmt.Errorf("unable to get repo %s: %w", r.GetFullName(), err)
+
+			// check if the retry limit has been exceeded
+			if i < retryLimit-1 {
+				logrus.WithError(err).Warningf("retrying #%d", i+1)
+
+				// continue to the next iteration of the loop
+				continue
+			}
+
+			return err
+		}
+
+		// update repo fields with any changes from SCM process
+		r.SetTopics(r.GetTopics())
+		r.SetBranch(r.GetBranch())
+
+		// set the build numbers based off repo counter
+		r.SetCounter(r.GetCounter() + 1)
```

Comment:
what are your thoughts on updating the repo counter in the database immediately after incrementing it, to reduce the risk of duplicate constrain violations? 

im just worried about compile + plan time for template-heavy pipelines and not making it to [this line](https://github.com/go-vela/server/pull/846/files#diff-6899fce6b909a4c271e1eabdda2f7240ad4ba8776c9a943e7b06b25d132e01f1R329) before an event-driven webhook is processed

Reply:
I'm uncertain we can do that since we'd be prematurely updating the `counter` for a repo before a build is triggered.

For the sake of the discussion, let's assume we do adopt this approach.

Let's say we're processing a schedule with the name of `X` for repo `Y` which has `10` builds.

As we're processing `X`, we'd increment the `counter` for `Y` from `10` -> `11` and update the DB.

Then, we attempt to compile the pipeline (let's pretend this works) for build number `11`.

Then, we attempt to add it to the DB (let's pretend this call fails):

https://github.com/go-vela/server/blob/6f740fa783dc2997ffe3617b553272007a3c951f/cmd/vela-server/schedule.go#L296-L307

From the above code, which was adopted from `webhook.go`, we'd retry which would go from `11` -> `12`.

And let's assume the second time we try all of this, everything works i.e. temp network failure

In this scenario, we'd never see a build for `11` since the second time we'd update the `counter` for `Y` to `12`.

Reply:
sure, then ignoring my incomplete solution, do you agree that the risk between this code and webhook events processing the same build number still exists? 

or am i thinking of it incorrect

Reply:
i suppose its in the same vein as the commented out TODO
```
		// TODO:
		// - if a build gets created and something else fails midway,
		//   the next loop will attempt to create the same build,
		//   using the same Number and thus create a constraint
		//   conflict; consider deleting the partially created
		//   build object in the database
```

the risk is small, and hopefully the retries would help.

Reply:
@plyr4 yeah I agree with you that the risk is still there 😞 

However, as you pointed out, the issue itself should hopefully be mitigated by retry logic.

FWIW, I did take the liberty of updating the number of retries from `3` -> `5` to help with this:

https://github.com/go-vela/server/blob/46fcc273b3aaea19f0e9fbc5a0db33212481c8f7/api/webhook.go#L414-L415

https://github.com/go-vela/server/blob/46fcc273b3aaea19f0e9fbc5a0db33212481c8f7/cmd/vela-server/schedule.go#L187-L188

To refresh folks memory on why we have it set to `3` in the webhook workflow:

https://github.com/go-vela/server/pull/191